Add one dimensional FFTs to torch.fft namespace

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44877 torch.fft helper functions
* #44876 Deprecate old fft functions
* #44550 torch.fft: Multi-dimensional transforms
* **#43011 Add one dimensional FFTs to torch.fft namespace**

Differential Revision: [D23751850](https://our.internmc.facebook.com/intern/diff/D23751850)

peterbell10 committed:
Add one dimensional FFTs to torch.fft namespace

[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 commented:
This is the first major step for #42175. NumPy compatible `fft`, `ifft`, `rfft`, `irfft` `hfft` and `ihfft`.

This still goes through the existing `_fft` helper function for ease of implementation and in order to maintain differentiability. But adds on top of it a number of NumPy/SciPy features like `dtype` promotion, sized FFTs, and a `dim` argument.

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

mruberry commented:
The backward_compatibility build failure is real:

```
Aug 13 16:08:16 Broken ops: [
Aug 13 16:08:16 	aten::_fft_with_size(Tensor self, int signal_ndim, bool complex_input, bool complex_output, bool inverse, int[] checked_signal_sizes, bool normalized, bool onesided, int[] output_sizes) -> (Tensor)
Aug 13 16:08:16 ]
```

I need to look at the PR proper to see if we can preserve it (maybe by wrapping it?). If not, this BC-break might be OK.

peterbell10 commented:
I am working on it. Wrapping the new signature should be fine.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,212 @@
+    return type;
+  }
+  // Promote integral to double
+  if (!at::isFloatingType(type)) {
```
For consistency with PyTorch should this instead promote to the default complex type?

peterbell10 replied:
Done.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,212 @@
+
+namespace {
+
+ScalarType promote_type(ScalarType type, bool require_complex) {
```
`promote_type` -> `promote_to_complex`? Something more precise would be nice.

peterbell10 replied:
The result can be a real dtype if `require_complex=False` is specified.

mruberry replied:
Good point. "promote_type_fft"? We already have "promote_types" which is a more general function with a very similar name.

peterbell10 replied:
Renamed to `promote_type_fft`.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,212 @@
+  TORCH_CHECK(false, "Invalid normalization mode: \"", *norm, "\"")
+}
+
+Tensor fix_shape(Tensor x, IntArrayRef dims, IntArrayRef sizes) {
```
Better add a comment explaining this function. Would something like `pad_fft` be a more descriptive name?

peterbell10 replied:
`pad` functions usually takes a padding amount relative to the current size, whereas this takes the desired shape (in the fft dimensions).

mruberry replied:
Good point. Maybe something more like "resize_fft_input"?

peterbell10 replied:
Renamed to `resize_fft_input`.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,212 @@
+  return must_copy ? at::constant_pad_nd(x, pad_amount) : x;
+}
+
+// Complex to real
```
This is a little misleading since there's already a complex to real transform in PyTorch, and this function will convert, pad, and manipulate its inputs, too.

peterbell10 replied:
Changed to "Complex to real FFT".

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,53 @@
 namespace torch {
 namespace fft {
 
-#ifndef DOXYGEN_SHOULD_SKIP_THIS
```
I'm not sure what the impact of removing the detail namespace is on doxygen. @glaringlee - do we need this pattern?

glaringlee replied:
@mruberry No, you don't need to follow the pattern. The DOXYGEN_SHOULD_SKIP_THIS is to prevent doxygen to generate the documents for its defined scope.  We don't want doxygen to generate documents for functions that defined in detail namespace, so we put there, this Macro is not related to any specific namespace.

mruberry replied:
Oh neat. So removing the indirection in these bindings is OK?

glaringlee replied:
Yes, I think so.

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,53 @@
+  return torch::fft_fft(self, n, axis, norm);
+}
+
+inline Tensor ifft(const Tensor& self,
```
Nit: simple doxygen comments for these, too. The torch:fft:fft documentation looks like this:

https://pytorch.org/cppdocs/api/function_namespacetorch_1_1fft_1ab483f290d94407a181ebb9793733d278.html?highlight=fft

Unfortunately it creates a link out of torch.fft.fft that doesn't work, though. @glaringlee maybe there's a better way to reference the Python documentation?

glaringlee replied:
@mruberry doxygen generates its documents within c++ domain. It can not ref python docs. We can manually add it in, but we don't have any way to add it automatically, they are total different domains. 
See this link:
https://pytorch.org/cppdocs/api/classtorch_1_1nn_1_1_multihead_attention_impl.html#classtorch_1_1nn_1_1_multihead_attention_impl
We put a doxygen style comments in MultiheadAttention impl, so it shows where is the python doc.

mruberry replied:
Got it. Thanks @glaringlee!

peterbell10 replied:
Added simple doxy-comments with a direct link to the python docs.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
 
 fft = _add_docstr(_fft.fft_fft, r"""
-fft(input) -> Tensor
+fft(input: Tensor, n: Optional[int] =None, dim: int =-1, norm: Optional[str] =None) -> Tensor
```
Despite the `-> Tensor` suggesting we type these strings (maybe we should?) we typically write them like normal Python signatures: `input, n=None, dim=-1, norm=None`

mruberry replied:
Would a note explaining how fft(real-valued tensor) is distinct from rfft be helpful?

peterbell10 replied:
Remove the type annotations.

peterbell10 replied:
>Would a note explaining how fft(real-valued tensor) is distinct from rfft be helpful?

There is a note to that effect in the `rfft` docs. Do you think it belongs here as well?

mruberry replied:
I noticed that. I would still include a note in fft, I think. Most people using the docs are probably just looking at one function returned by the search feature, so the note in rfft will be hard to discover for people looking at fft, and I think the interpretation by fft of non-complex dtype inputs as complex dtype inputs is not immediately obvious.

peterbell10 replied:
Added a note about hermitian symmetry and `rfft` to the `fft` doc.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+  n (int, optional): FFT length. If given, the input will either be zero-padded
+    or trimmed to this length before taking the FFT.
+  dim (int, optional): The dimension along which to take the one dimensional FFT.
+  norm (str, optional): Normalization mode:
```
In NumPy norm is None or "ortho." Having a superset (or subset) of NumPy functionality is OK, but I'm confused by the "forward" and "backward" here since "backward" says it's normalizing `torch.fft.ifft`. Can this clarify what's going on?

Also, the default from the signature is None, not "backward." This could say "if norm is None then 'backward' normalization is used?"

peterbell10 replied:
These will be in the next NumPy and SciPy releases. See numpy/numpy#16476 and scipy/scipy#12532.

peterbell10 replied:
And `None` is the same as `"backward"` which is why that is given as the default. This is consistent with the numpy docstring.

mruberry replied:
That's cool. Maybe it's just confusing to me? I still think we could improve the clarity of what's happening here for readers.

peterbell10 replied:
Reworded to "the {forward,backward} transform ({fft,ifft}) is normalized by `1/n`. Hopefully that is clearer.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+  dim (int, optional): The dimension along which to take the one dimensional IFFT.
+  norm (str, optional): Normalization mode:
+    ``"forward"`` to normalize the :func:`torch.fft.fft` by ``1/n``,
+    ``"backward"`` to normalize the :func:`torch.fft.ifft` by ``1/n`` or
```
Same question here.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
 
-    >>> t = torch.randn(4, dtype=torch.complex128)
+    >>> import torch.fft
+    >>> t = torch.arange(4)
```
+ an interesting complex example to highlight c2c functionality?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+""")
+
+ifft = _add_docstr(_fft.fft_ifft, r"""
+ifft(input: Tensor, n: Optional[int] =None, dim: int =-1, norm: Optional[str] =None) -> Tensor
```
Same typing comment as above

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+    ``"forward"`` to normalize the :func:`torch.fft.fft` by ``1/n``,
+    ``"backward"`` to normalize the :func:`torch.fft.ifft` by ``1/n`` or
+    ``"ortho"`` to normalize both by ``1/sqrt(n)`` (making it orthonormal).
+    Default, is ``"backward"``
```
Same default note here

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+    >>> import torch.fft
+    >>> t = torch.arange(4)
+    >>> t
+    tensor([0, 1, 2, 3])
```
Maybe a c2c example that pairs with the fft example above?

mruberry replied:
(e.g. a nontrivial version of `np.fft.ifft(np.fft.fft(a))`?)

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+""")
+
+rfft = _add_docstr(_fft.fft_rfft, r"""
+rfft(input: Tensor, n: Optional[int] =None, dim: int =-1, norm: Optional[str] =None) -> Tensor
```
see above

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,227 @@
+  dim (int, optional): The dimension along which to take the one dimensional IFFT.
+  norm (str, optional): Normalization mode:
+    ``"forward"`` to normalize the :func:`torch.fft.rfft` by ``1/n``,
+    ``"backward"`` to normalize the :func:`torch.fft.irfft` by ``1/n`` or
```
per above

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

mruberry commented on test/cpp/api/fft.cpp:
```diff
@@ -5,12 +5,123 @@
 
 #include <test/cpp/api/support.h>
 
+// Naive DFT of a 1 dimensional tensor
```
These tests are super cool and we should definitely keep them, but in general let's bias towards Python testing where it's possible and reserve future C++ tests for testing what we can't test from Python.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -5,11 +5,15 @@
 
 from torch.testing._internal.common_utils import \
-    (TestCase, run_tests, TEST_NUMPY, TEST_LIBROSA)
+    (TestCase, run_tests, TEST_NUMPY, TEST_LIBROSA, _assertGradAndGradgradChecks)
```
For autograd checks update method_tests in torch/testing/_internal/common_methods_invocations.py. 

I'm not sure if method_tests requires tests have method variants (it shouldn't), but if it does you can add method variants for these functions and not document them.

peterbell10 replied:
I added a simple test to `common_method_invocations.py`. I also had to modify `test_autograd.py` to support qualified names like `'fft.fft'`.

mruberry replied:
You could also use the ugly torch name, like torch.fft_fft.

peterbell10 replied:
That doesn't exis in the torch namespace, it's only in `torch._C._fft.fft_fft`.

mruberry replied:
Oh you're right, sorry. It's only in the C++ namespace that we still have names like fft_fft.

mruberry commented:
Also needs to update https://github.com/pytorch/pytorch/blob/master/docs/source/fft.rst to link the doc strings.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,162 @@ def method_fn(t):
-                self.assertEqual(actual, expected, exact_dtype=(dtype is torch.complex128))
+    @precisionOverride({torch.complex64: 1e-4, torch.float: 1e-4})
+    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
+    def test_fft_numpy(self, device, dtype):
```
This test is really excellent.

mruberry replied:
Overall these tests are very, very good. Is it interesting, after checking for consistency with our NumPy reference, to also check for stuff like `x = ifft(fft(x))`?

Also, are there any special cases or oddities we should be checking, like an fft formulation with -infs, infs, or NaNs?

peterbell10 replied:
Added round-trip tests. Definitely makes sense so we can have a test that works independent of numpy.

>are there any special cases or oddities we should be checking, like an fft formulation with -infs, infs, or NaNs?

Not that I'm aware of. It's pretty much entirely adds and multiplications so should propagate nans and infs correctly by normal IEEE semantics.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,162 @@ def method_fn(t):
-
-        # CPU (MKL)
-        with self.assertRaisesRegex(RuntimeError, "MKL FFT error"):
+        with self.assertRaisesRegex(RuntimeError, match):
```
Nit: you can code golf this by iterating over the ops:

```
for op in (torch.fft.fft, torch.fft.ifft, ...):
```

You'd have to expand the op list based on the dtype like you do in the previous test to handle rfft and ihfft.

peterbell10 replied:
Done.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,162 @@ def method_fn(t):
-        with self.assertRaisesRegex(RuntimeError, "Expected a complex tensor"):
-            torch.fft.fft(t)
+        PROMOTION_MAP = {
+            torch.int8: torch.complex64,
```
This may need a tweak depending on what we decided for int promotions (per above comment).

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,162 @@ def method_fn(t):
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @dtypes(torch.double, torch.complex128)  # gradcheck requires double
+    def test_fft_backward(self, device, dtype):
```
This is cool and I like checking gradients directly in these "vertical" test suites. Maybe we should still add method_test entries for the moment?

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 commented:
@mruberry this is ready for another review.

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+    type = c10::typeMetaToScalarType(c10::get_default_dtype());
+  }
+
+  TORCH_CHECK(type == kFloat || type == kDouble, "Unsupported dtype ", type);
```
correct me if I am wrong but anytime the input type is not complex, it will just error out, right? if so, the code after this check will never be executed ...

anjali411 commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+    type = c10::typeMetaToScalarType(c10::get_default_dtype());
+  }
+
+  TORCH_CHECK(type == kFloat || type == kDouble, "Unsupported dtype ", type);
```
so the only case this will error out is if the default dtype is set to torch.half?

peterbell10 replied:
`half` or `bfloat16`.

anjali411 commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+Tensor promote_tensor_fft(const Tensor& t, bool require_complex=false) {
+  auto cur_type = t.scalar_type();
+  auto new_type = promote_type_fft(cur_type, require_complex);
+  return (cur_type == new_type) ? t : t.to(new_type);
```
Currently float tensors (containing complex data) input to fft have shape `(...,2)`, so would `t.to(new_type)` suffice for cases like `promote_type_fft(kFloat, True)`? wouldn't you wanna use `torch.view_as_complex` in such cases?

peterbell10 replied:
These functions do not treat a `(..., 2)` real tensor as though it were complex. This takes real input and promotes it to a complex tensor with zero imaginary component, regardless of shape.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
 #include <cmath>
 
 namespace at { namespace native {
+static inline Tensor _fft(
```
newline + comment

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+
+    if (x_sizes[dims[i]] < sizes[i]) {
+      must_copy = true;
+      auto pad_idx = pad_amount.size() - 2 * dims[i] - 1;
```
So if dims is [0, 1] and `pad_amount.size()` == 4, this could access indices 3 and 1?

peterbell10 replied:
That's right. Padding is always applied to the right side so the index is always odd, and `pad`'s list starts at the last dimension and goes backwards.

https://github.com/pytorch/pytorch/blob/e37a24b0446e80d01b52cb04a2e7fde6976a16d4/aten/src/ATen/native/ConstantPadNd.cpp#L49

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+    }
+  }
+
+  // Only call pad if needed, otherwise we copy the entire tensor
```
This comment seems misleading because it's "pad or copy," but the ternary operator is "if must copy pad, else return"

peterbell10 replied:
Rephrased to be clear that `pad` would clone.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+  input = promote_tensor_fft(input, /*require_complex=*/true);
+  const auto input_dim = input.dim();
+  const auto dim = maybe_wrap_dim(unwrapped_dim, input_dim);
+  const auto n = n_opt.value_or(input.sizes()[dim] * 2 - 1);
```
This is interesting. I would have expected the default for `n` to be the size of the dim. Why the *2?

peterbell10 replied:
`n` refers to the size of the full signal before dropping hermetian symmetric terms.
See https://numpy.org/doc/stable/reference/generated/numpy.fft.irfft.html

Although now I look again, this should actually be `2 * (input.sizes()[dim] - 1)`.

peterbell10 replied:
I was concerned why this wasn't picked up by the testing. It seems you can't iterate over an `itertools` iterator multiple times, so I'm making `test_args` a list instead.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+  if (n_opt) {
+    input = resize_fft_input(input, dim, n/2 + 1);
+  }
+  const bool must_transpose = (dim != input_dim - 1);
```
Better add a comment for why this transform is needed

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+  }
+  auto out = _fft(at::view_as_real(input),
+                  /*signal_ndim=*/1, /*complex_input=*/true,
+                  /*complex_output=*/false, /*inverse=*/true, {n},
```
`/*signal_sizes=*/n`?

mruberry commented on aten/src/ATen/native/SpectralOpsUtils.h:
```diff
@@ -6,6 +6,12 @@
 namespace at { namespace native {
 
+enum class fft_norm_mode {
+  none,
```
comment for enum class + comment for each type

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+
+namespace {
+
+ScalarType promote_type_fft(ScalarType type, bool require_complex) {
```
comment

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+  TORCH_INTERNAL_ASSERT(false, "Unhandled dtype");
+}
+
+Tensor promote_tensor_fft(const Tensor& t, bool require_complex=false) {
```
comment here

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -17,23 +17,221 @@
+  return (cur_type == new_type) ? t : t.to(new_type);
+}
+
+fft_norm_mode norm_from_string(c10::optional<std::string> norm, bool forward) {
```
comment here

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,208 @@ def method_fn(t):
+            (-1, 0),
+            # norm
+            ((None, "forward", "backward", "ortho")
+             if LooseVersion(np.__version__) >= '1.20.0'
```
Cool.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    dim (int, optional): The dimension along which to take the one dimensional FFT.
+    norm (str, optional): Normalization mode:
+        ``"forward"`` - the forward transform (:func:`~torch.fft.fft`) is normalized by ``1/n``,
+        ``"backward"`` - the backward transform (:func:`~torch.fft.ifft`) is normalized by ``1/n``,
```
So what does calling `fft` with norm mode backward actually do? These statements are clearer but it's still odd to me that the backward description refers to a different function.

peterbell10 replied:
If you do `ifft(fft(x, mode=mode), mode=mode)` then no matter what mode you use, the total normalizing factor of `1/N` is applied. What changes is which transform does the normalization. So, `fft(x, norm="backward")` is the unnormalized FFT and `ifft(x, norm="forward")` is also unnormalized.

mruberry replied:
I think I get that and I understand where the documentation is coming from, but maybe the "backward" description could refer to its effect on the function being documented (unnormalized) and then add a follow-up sentence to include this perspective? Because as it is the effect of "backward" on fft is unspecified.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Computes the one dimensional discrete Fourier transform of real :attr:`input`.
+
+The FFT of a real signal is hermitian-symmetric, ``X[i] = conj(X[-i])`` so
+the output contains only the positive frequencies below the nyquist frequency.
```
Nyquist (proper name)

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    tensor([ 6.+0.j, -2.+2.j, -2.+0.j, -2.-2.j])
+
+    Notice that the symmetric element ``T[-1] == T[1].conj()`` is ommitted.
+    At the Nyquist freqency ``T[-2] == T[2].conj() == T[2]`` is given but must
```
freqency -> frequency

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
-    tensor([-2.5086+0.9281j, -0.0646+0.8343j, -1.7846-2.9435j, -0.1878-1.0965j], dtype=torch.complex128)
+    tensor([ 6.+0.j, -2.+2.j, -2.+0.j, -2.-2.j])
+
+    Notice that the symmetric element ``T[-1] == T[1].conj()`` is ommitted.
```
ommitted -> omitted

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Computes the inverse of :func:`torch.fft.rfft`.
+
+:attr:`input` is a one-sided hermitian signal in the fourrier domain, as
```
fourrier -> fourier

mruberry replied:
hermitian -> Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Computes the inverse of :func:`torch.fft.rfft`.
+
+:attr:`input` is a one-sided hermitian signal in the fourrier domain, as
+produced by :func:`torch.fft.rfft`. By the hermitian property, the time-domain
```
hermitian -> Hermitian

mruberry replied:
(weird that it's not the "Hermite property," right?)

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+produced by :func:`torch.fft.rfft`. By the hermitian property, the time-domain
+signal will be purely real.
+
+Note that some input frequencies must be purely real to satisfy the hermitian
```
hermitian -> Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+property. In these cases the imaginary component will be ignored.
+
+Args:
+    input (Tensor): the input tensor representing a half-hermitian signal
```
half-hermitian -half-Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    >>> T
+    tensor([ 6.+0.j, -2.+2.j, -2.+0.j])
+
+    Without specifying the output length to :func:`torch.fft.irfft`, the output
```
Should this be a note and then the example can have a comment like "passing signal length to support round-trips"?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+hfft = _add_docstr(_fft.fft_hfft, r"""
+hfft(input, n=None, dim=-1, norm=None) -> Tensor
+
+Computes the one dimensional discrete Fourier transform of a hermitian
```
Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Computes the one dimensional discrete Fourier transform of a hermitian
+symmetric :attr:`input` signal.
+
+Because the signal is hermitian in the time-domain, the result will be real in
```
Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Because the signal is hermitian in the time-domain, the result will be real in
+the frequency domain. Note that some input frequencies must be purely real to
+satisfy the hermitian property. In these cases the imaginary component will be
```
Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+ignored.
+
+Args:
+    input (Tensor): the input tensor representing a half-hermitian signal
```
half-Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Example::
+
+    Taking a frequency signal that's purely real and bringing into the time
+    domain gives hermitian symmetric output:
```
Hermitian

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Example::
+
+    Taking a frequency signal that's purely real and bringing into the time
```
bringing into -> bringing it into

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Example::
+
+    Taking a frequency signal that's purely real and bringing into the time
```
If text is in an example it might render with code formatting, which could make it tricky to read. You could put the text inline like the other documentation text and then do `Example::` for the code snippets (possibly with comments). 

I could also be mistaken. If you try rendering this maybe it will render correctly.

peterbell10 replied:
I built the docs, and removing the second colon made the text render as text and the doctest style code is still rendered as code.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Computes the inverse one dimensional Fourier transform of a real signal.
+
+The IFFT of a real signal is hermitian-symmetric, ``X[i] = conj(X[-i])`` so
```
Hermitian-symmetric

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Computes the inverse one dimensional Fourier transform of a real signal.
+
+The IFFT of a real signal is hermitian-symmetric, ``X[i] = conj(X[-i])`` so
+the output contains only the positive frequencies below the nyquist frequency.
```
Nyquist frequency

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    input (Tensor): the real input tensor
+    n (int, optional): FFT length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
+    dim (int, optional): The dimension along which to take the one dimensional IFFT.
```
IFFT or IHFFT?

peterbell10 replied:
Changed all of these to either real (I)FFT or Hermitian (I)FFT.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Args:
+    input (Tensor): the real input tensor
+    n (int, optional): FFT length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
```
FFT or IHFFT?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    n (int, optional): Output signal length. This determines the length of the
+        real output. If given, the input will either be zero-padded or trimmed to this
+        length before taking the FFT. Defaults to odd output: ``n=input.size(dim) * 2 - 1``.
+    dim (int, optional): The dimension along which to take the one dimensional FFT.
```
FFT or HFFT?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    n (int, optional): Output signal length. This determines the length of the
+        output signal. If given, the input will either be zero-padded or trimmed to this
+        length before taking the IFFT. Defaults to odd output: ``n=input.size(dim) * 2 - 1``.
+    dim (int, optional): The dimension along which to take the one dimensional IFFT.
```
IFFT or RFFT?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+    input (Tensor): the real input tensor
+    n (int, optional): FFT length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
+    dim (int, optional): The dimension along which to take the one dimensional IFFT.
```
IFFT or RFFT?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+
+Computes the one dimensional discrete Fourier transform of real :attr:`input`.
+
+The FFT of a real signal is hermitian-symmetric, ``X[i] = conj(X[-i])`` so
```
Hermitian-symmetric

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,233 @@
+Args:
+    input (Tensor): the input tensor
+    n (int, optional): FFT length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
```
FFT or IFFT?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1276,6 +1282,12 @@ def unpack_variables(args):
 EXCLUDE_GRADCHECK = {
 }
 EXCLUDE_GRADGRADCHECK = {
+    'fft.fft',
```
No double backward support is OK - what's the blocker on double backward support, though?

peterbell10 replied:
I believe it's `_fft_with_size` that's failing `gradgradcheck`. I assume it just wasn't written with gradgrad in mind. I can revisit this later; I don't think there's a fundamental reason it couldn't work.

mruberry reviewed (changes requested):
Overall continues to look very good, @peterbell10. Sorry to make you wait for a review. The next one should be very fast since at this point the code and tests look mostly excellent. Most of my comments are nits about commenting the code and doc issues, plus a few questions. One those are addressed I think we're good to go!

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,208 @@ def method_fn(t):
+            # n
+            (None, 50, 6),
+            # dim
+            (-1, 0),
```
Are there ever tests for an intermediate dim, like a 3D tensor with dim = 1?

peterbell10 replied:
Added another set of arguments for middle dimensions.

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -18,22 +18,235 @@
+  return (cur_type == new_type) ? t : t.to(new_type);
+}
+
+// Convert numpy-compatible normalization mode string into internal normalization mode enum
```
nit: "Converts NumPy compatible normalization mode strings to enum values"

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -18,22 +18,235 @@
+}
+
+// Convert numpy-compatible normalization mode string into internal normalization mode enum
+// Numpy modes change meaning between the FFT and IFFT, so the transform direction must also be given.
```
nit: "NOTE: NumPy's normalization modes have direction-specific meanings. For example, ... "

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -18,22 +18,235 @@
+}
+
+// Fixes the shape of x such that x.size(dims[i]) == sizes[i]
+// Either by zero-padding, or by slicing x starting from 0
```
nit: comma at end of previous line, start this with lowercase "either" since it's not a new sentence

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -18,22 +18,235 @@
+    }
+  }
+
+  // Only call pad if needed, otherwise pad will clone the entire tensor
```
Nit: "Only calls pad if necessary since pad copies the entire tensor"

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -139,6 +354,25 @@ static inline Tensor _fft(const Tensor &self, const int64_t signal_ndim,
 }
 
+// Wrapper to preserve the historic signature of _fft_with_size
+// This is only used for torch-script back-compat and the new signature with
```
nit: "NOTE: This is only used for torchscript backwards compatibility, and the new signature..."

mruberry commented on test/test_autograd.py:
```diff
@@ -4668,9 +4685,10 @@ def fn(*inputs):
+                    if torch_fn is not None and name not in EXCLUDE_FUNCTIONAL:
                         def fn(*inputs):
-                            output = getattr(torch, name)(*inputs, **kwargs)
+                            output = getattr_qualified(torch, name)(*inputs, **kwargs)
```
can use torch_fn here instead of calling getattr_qualified again?

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,113 @@
-namespace detail {
+/// Computes the 1 dimensional fast fourier transform over a given axis
+///
+/// \see https://pytorch.org/docs/master/fft.html#torch.fft.fft
```
nit: "Computes the one-dimensional fast Fourier transform over a given axis. See <url>."

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,113 @@
+  return torch::fft_ifft(self, n, axis, norm);
+}
+
+/// Computes the 1 dimensional FFT of real input with half-complex output
```
What does "half-complex output" mean?

mruberry replied:
Same question for the following docs, too.

peterbell10 replied:
The compact Hermitian representation is complex but only has half the number of values, hence "half-complex". Maybe you would prefer "one-sided"?

mruberry replied:
One-sided sounds better. "half-complex" makes me think only the real or imaginary parts are present (i.e. only half of each complex value).

peterbell10 replied:
Changed here and tried to use onesided consistently in the python docs as well.

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,113 @@
-  return torch::fft_fft(self);
+/// Computes the inverse of torch.fft.rfft
+///
+/// The input is a half-complex fourier domain signal, with purely real output.
```
instead of "purely real" maybe just "real-valued"?

peterbell10 replied:
Changed everywhere.

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -5,25 +5,113 @@
+/// Computes the 1 dimensional FFT of a half-complex signal
+///
+/// The half-complex input represents a hermitian symmetric time domain signal.
+/// The returned fourier domain representation of such a signal is a purely
```
"is a purely real.' -> "... is real-valued."

mruberry commented on torch/csrc/api/include/torch/fft.h:
````diff
@@ -5,25 +5,113 @@
+/// ```
+/// auto T = torch::randn(128 / 2 + 1, torch::kComplexDouble);
+/// auto t = torch::fft::irfft(t, /*n=*/128);
+/// assert(!t.is_complex() && T.numel() == 128);
````
Maybe a positive assertion like `t.is_floating_point()` would make the description clearer? Same comment below.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    Hermitian property: `X[i] = conj(X[-i])`. This function always returns both
+    the positive and negative frequency terms, even though for real inputs the
+    negative frequencies are redundant. :func:`~torch.fft.rfft` returns the more
+    efficient "half-complex" representation where only the positive frequencies
```
This is a reasonable description of "half-complex," but readers of the C++ docstrings may not encounter it.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
-  input (Tensor): the input tensor
+    input (Tensor): the input tensor
+    n (int, optional): Signal length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
```
"before taking" -> "before computing"

peterbell10 replied:
changed everywhere.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    n (int, optional): Signal length. If given, the input will either be zero-padded
+        or trimmed to this length before taking the FFT.
+    dim (int, optional): The dimension along which to take the one dimensional FFT.
+    norm (str, optional): Normalization mode. Between the forward and backward
```
This is much better and I think it's already reasonably clear, but I'd still like to try and tweak it once more because this documentation assumes the reader intends to compute the fft and then invert it.  What if we structured this as follows:

- there are 3 normalization modes, forward, backward, and ortho
(bullet points)
- forward normalization mode will cause fft to normalize by 1/n
- backward normalization mode will cause fft to apply no normalization
- ortho will cause fft to apply the 1/sqrt(n) normalization
(end of bullet points)
- the default is "backward" normalization (no normalization)
- These normalization modes correspond with the normalization modes for ifft. For example, the "backward" normalization mode causes fft, the forward function, to apply no normalization, but "backward" normalization mode causes ifft, the backward function, to normalize by 1/n. This means that if the inverse function is applied to the output of the fft function and both have the same normalization mode then a normalization factor of 1/n will have been applied.

That's not perfect as written, but the idea is to capture just the docs for fft devoid of context, then add that perspective. Essentially it's reversing the order of what you've written. 

What are your thoughts? Is it worth trying to tweak this a little?

peterbell10 replied:
I'm a bit reluctant because it doesn't relate to the mode names very well. How about this draft:

```
norm (str, optional): Normalization mode. For the forward transform (:func:`~torch.fft.fft`),
    these correspond to:

    * ``"forward"`` - normalize by ``1/n``
    * ``"backward"`` - no normalization
    * ``"ortho"`` - normalize by ``1/sqrt(n)`` (making the FFT orthonormal)

    Calling the backward transform (:func:`~torch.fft.ifft`) with the same
    normalization mode will apply an overall normalization of ``1/n`` between
    the two transforms. This is required to makes `ifft` the exact inverse.
```

mruberry replied:
Yes your suggestion is succinct, clear, and much better. Do you want to still put something like, "The default is 'backward' (no normalization)." at the end?

peterbell10 replied:
I've updated all the docs to this and left the default line as it was.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+        * ``"backward"`` - the backward transform is normalized by ``1/n``, forward is unnormalized.
+        * ``"ortho"`` - both transforms are normalized by ``1/sqrt(n)`` (making it orthonormal).
+
+        Default is ``"backward"``.
```
nit: Default is backward (normalize by 1/n). 

If you like the suggestion above about how to explain these modes you could also point to the documentation for fft to explain the normalization mode relationship instead of duplicating the description.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+rfft = _add_docstr(_fft.fft_rfft, r"""
+rfft(input, n=None, dim=-1, norm=None) -> Tensor
+
+Computes the one dimensional discrete Fourier transform of real :attr:`input`.
```
"real" -> "real-valued"

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    tensor([ 6.+0.j, -2.+2.j, -2.+0.j, -2.-2.j])
+
+    Notice that the symmetric element ``T[-1] == T[1].conj()`` is omitted.
+    At the Nyquist frequency ``T[-2] == T[2].conj() == T[2]`` is given but must
```
I'm not sure what this last statement is trying to say.

peterbell10 replied:
Rewritten as
>    At the Nyquist frequency ``T[-2] == T[2]`` is it's own symmetric pair,
    and therefore must always be real-valued.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    >>> torch.fft.rfft(t)
+    tensor([ 6.+0.j, -2.+2.j, -2.+0.j])
+
+    Compare against the full output from :func:`~torch.fft.fft`:
```
This is a really cool example.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+signal will be purely real.
+
+Note:
+    Some input frequencies must be purely real to satisfy the Hermitian
```
instead of "purely real" maybe "real-valued" (if float) or "have no imaginary part" or "have an imaginary part of zero" (if complex)?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+
+Computes the inverse of :func:`~torch.fft.rfft`.
+
+:attr:`input` is a one-sided Hermitian signal in the Fourier domain, as
```
Nit: this is OK but technically the function is "interpreting" input this way or we should say that input is "typically" a one-sided Hermitian signal or "expected to be". That is, this description conflates the function's intentionality (what it's intended to do) with the actual mechanics of the function (which compute something even if it's given random values devoid of mathematical perspective).

peterbell10 replied:
Changed to "is interpreted as a one-sided Hermitian signal".

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+Computes the inverse of :func:`~torch.fft.rfft`.
+
+:attr:`input` is a one-sided Hermitian signal in the Fourier domain, as
+produced by :func:`~torch.fft.rfft`. By the Hermitian property, the time-domain
```
I'm not sure what this second sentence is getting at. Is it trying to constrain the domain of the function?

peterbell10 replied:
This is describing the output. The input is Hermitian, therefore the output is real. Perhaps it would be clearer to say "the output will be real-valued"?

peterbell10 replied:
Changed.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+
+Note:
+    Some input frequencies must be purely real to satisfy the Hermitian
+    property. In these cases the imaginary component will be ignored.
```
I'm a little confused how this sentence relates to the previous one. 

This function is interpreting input as a one-sided Hermitian signal in the Fourier domain. Is this saying that if that input doesn't satisfy an expected invariant then it will be interpreted as if it did?

peterbell10 replied:
Yes, for example if the the zero-frequency term had an imaginary component then it would be impossible to include that imaginary term in the real output. So it would always just be ignored.

mruberry replied:
Got it. What about adding a brief, "For example, ..." sentence after at the end of this to clarify just like you did here?

peterbell10 replied:
Added the example to this and `hfft`.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    property. In these cases the imaginary component will be ignored.
+
+Note:
+    The correct interpretation of the hermitian input depends on the length of
```
"hermitian" -> "Hermitian"

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+        length before taking the real IFFT.
+        Defaults to even output: ``n=2*(input.size(dim) - 1)``.
+    dim (int, optional): The dimension along which to take the one dimensional real IFFT.
+    norm (str, optional): Normalization mode. Between the forward and backward
```
The more I read this normalization mode description the more I appreciate it, but I still want to try to tweak it slightly per the above.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+    >>> torch.fft.irfft(T)
+    tensor([0.6250, 1.4045, 3.1250, 4.8455])
+
+    So, it is recommended to always pass the signal length :attr:`n`:
```
Cool example.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+symmetric :attr:`input` signal.
+
+Because the signal is Hermitian in the time-domain, the result will be real in
+the frequency domain. Note that some input frequencies must be purely real to
```
Note should probably be a "Note:".

I have the same question here as above about ignoring the imaginary component. One way to deal with this might be to include an example, too.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+ignored.
+
+Note:
+    The correct interpretation of the hermitian input depends on the length of
```
"hermitian" -> "Hermitian"

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+satisfy the Hermitian property. In these cases the imaginary component will be
+ignored.
+
+Note:
```
These notes are really good.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+            -0.5000+0.6882j])
+
+    Note that ``T[1] == T[-1].conj()`` and ``T[2] == T[-2].conj()`` is
+    redundant. We can thus compute the forward transform without considering
```
Excellent example.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+ihfft = _add_docstr(_fft.fft_ihfft, r"""
+ihfft(input, n=None, dim=-1, norm=None) -> Tensor
+
+Computes the inverse one dimensional Fourier transform of a real signal.
```
This description doesn't seem correct and the following paragraph mentions IFFT.

peterbell10 replied:
Changed this line to by "the inverse of hfft" and tweaked the wording of the next paragraph slightly to emphasize `ihfft` is onesided.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -9,18 +9,283 @@
+
+Args:
+    input (Tensor): the input tensor representing a half-Hermitian signal
+    n (int, optional): Output signal length. This determines the length of the
```
Totally optional idea is to include, as a note, a comment like the NumPy documentation for hfft has explaining how hfft and ihfft are analogous to rfft and irfft but "for the opposite case," which is why hfft has to specify the output signal length.

mruberry reviewed (commented):
Hey @peterbell10, thanks for this update. 

I've made a lot of comments and suggestions for the documentation. Writing is really hard, and I want to be clear that overall the documentation is already very good. 

What I'd like to propose we do, however, is make one more update based on this feedback, then call the PR "good enough" and leave remaining documentation ideas/improvements/suggestions as future work. This PR is massive and contains a lot of excellent work, and I don't think we should block it while we take the documentation from A to an A+, for example. 

Does that sound reasonable to you?

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 commented:
>Does that sound reasonable to you?

Absolutely. I'd like to get all the `torch.fft` functionality in for the 1.7 branch point and once the functionality is there, I can put in any finishing touches that are needed.

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"




[ghstack-poisoned]

peterbell10 commented on test/test_spectral_ops.py:
```diff
@@ -43,57 +47,229 @@ def method_fn(t):
+                input = args[0]
+                args = args[1:]
+
+                # Workaround for gradcheck's poor support for complex input
+                # Use real input instead and put view_as_complex into the graph
+                if dtype.is_complex:
+                    def test_fn(x):
+                        return torch_fn(torch.view_as_complex(x), *args)
+                    input = torch.view_as_real(input).detach().requires_grad_()
+                else:
+                    def test_fn(x):
+                        return torch_fn(x, *args)
+                    input = input.detach().requires_grad_()
+
+                self.assertTrue(torch.autograd.gradcheck(test_fn, (input,)))
```
Just to draw attention to this. I was getting `gradcheck` failures for `irfft` but when I looked into it, the numerical jacobian calculated inside of `gradcheck` was actually wrong and tweaking the `eps` didn't help. `torch.irfft` however worked fine, and I found that making the leaf node purely real fixed the numerical jacobian issue.

mruberry replied:
Interesting. cc @albanD and @anjali411, who are working on complex autograd.

albanD replied:
Yes there is not support for complex in gradcheck yet. So only running it in the real case is enough right now.

mruberry commented:
Hey @peterbell10, ping me when you think this is ready.

peterbell10 commented:
@mruberry This should be ready now, if you're okay with the `gradcheck` workaround.

mruberry reviewed (approved):
Awesome work, @peterbell10! This PR is more like three challenging PRs combined into one. Thank you for your diligence and attention to detail.

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"


Differential Revision: [D23751850](https://our.internmc.facebook.com/intern/diff/D23751850)

[ghstack-poisoned]

mruberry commented:
Unfortunately it looks like this is causing an internal testing issue where our test runner is trying but failing to extract test_fft from TestJitGeneratedAutogradCPU. I'm not entirely sure what's going on here. I think one simple fix would be to delete the fft entries from `method_tests` in common_methods_invocations, since they're what's used to generated these tests. That will reduce the test surface but not in a significant way, and in the future we're moving away from method_tests and towards OpInfos, anyway.

Would you mind removing those tests, @peterbell10?

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"


Differential Revision: [D23751850](https://our.internmc.facebook.com/intern/diff/D23751850)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add one dimensional FFTs to torch.fft namespace"


Differential Revision: [D23751850](https://our.internmc.facebook.com/intern/diff/D23751850)

[ghstack-poisoned]

peterbell10 commented:
Removed the `common_methods_invocations` stuff and rebased on master. @mruberry
